extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/CatalogUpdateNotifier.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.druid.catalog.http.CatalogListenerResource;
+import org.apache.druid.catalog.model.TableId;
+import org.apache.druid.catalog.model.TableMetadata;
+import org.apache.druid.catalog.model.TableSpec;
+import org.apache.druid.catalog.storage.CatalogStorage;
+import org.apache.druid.catalog.sync.MetadataCatalog.CatalogListener;
+import org.apache.druid.catalog.sync.RestUpdateSender.RestSender;
+import org.apache.druid.discovery.DruidNodeDiscoveryProvider;
+import org.apache.druid.discovery.NodeRole;
+import org.apache.druid.guice.ManageLifecycle;
+import org.apache.druid.guice.annotations.EscalatedClient;
+import org.apache.druid.guice.annotations.Smile;
+import org.apache.druid.java.util.common.lifecycle.LifecycleStart;
+import org.apache.druid.java.util.common.lifecycle.LifecycleStop;
+import org.apache.druid.java.util.http.client.HttpClient;
+import org.apache.druid.server.DruidNode;
+import org.joda.time.Duration;
+
+import javax.inject.Inject;
+
+import java.util.Collections;
+import java.util.function.Supplier;
+
+/**
+ * Global update notifier for the catalog. Registers itself as a catalog
+ * listener, then uses the common cache notifier to send Smile-encoded JSON
+ * updates to broker nodes discovered from node discovery (typically ZooKeeper.)
+ * <p>
+ * Deletes are encoded as a table update with a table definition of a special
+ * tombstone type. This saves having the need for two endpoints, or having
+ * a wrapper class to handle deletes.
+ */
+@ManageLifecycle
+public class CatalogUpdateNotifier implements CatalogListener
+{
+  private static final String CALLER_NAME = "Catalog Sync";
+  private static final long TIMEOUT_MS = 5000;
+
+  /**
+   * Internal table type used in updates to notify listeners that a table has
+   * been deleted. Avoids the need for a special "table deleted" message.
+   */
+  public static final String TOMBSTONE_TABLE_TYPE = "tombstone";
+  private static final TableSpec TABLE_TOMBSTONE = new TableSpec(TOMBSTONE_TABLE_TYPE, null, null);
+
+  private final CacheNotifier notifier;
+  private final ObjectMapper smileMapper;
+
+  @Inject
+  public CatalogUpdateNotifier(
+      CatalogStorage catalog,
+      DruidNodeDiscoveryProvider discoveryProvider,
+      @EscalatedClient HttpClient httpClient,
+      @Smile ObjectMapper smileMapper
+  )
+  {
+    long timeoutMs = TIMEOUT_MS;
+    this.smileMapper = smileMapper;
+    Supplier<Iterable<DruidNode>> nodeSupplier = new ListeningNodeSupplier(
+        Collections.singletonList(NodeRole.BROKER),
+        discoveryProvider);
+    RestSender restSender = RestUpdateSender.httpClientSender(httpClient, Duration.millis(timeoutMs));
+    RestUpdateSender sender = new RestUpdateSender(
+        CALLER_NAME,
+        nodeSupplier,
+        restSender,
+        CatalogListenerResource.BASE_URL + CatalogListenerResource.SYNC_URL,
+        timeoutMs);
+    this.notifier = new CacheNotifier(
+        CALLER_NAME,
+        sender);
+    catalog.register(this);
```

Comment:
Should only the leader be able to register itself and push updates to listeners? If so, what is controlling that?

Reply:
How do we handle this issue generically? I'd kind of hoped that the Coordinator itself will reject messages if it is the follower.

Reply:
Take a look at the `TaskMaster` class, you'll see that it does things like startup the taskRunner on on the leader when it comes up, and shutdown on follower when is loses leadership. Maybe can follow that sort of pattern? Not sure if needed. There is logic that routes the coordinator request to leader, but not sure if there is anyway for that to get out of sync with the leadership election, and route request to wrong host.

Reply:
Yeah maybe this isnt needed. Will just leave up here, but feel free to resolve, if you feel we dont need too.